R/module_teal.R

```diff
@@ -184,33 +186,91 @@ srv_teal <- function(id, data, modules, filter = teal_slices()) {
       }
     )
 
-    data_rv <- srv_init_data("data", data = data, modules = modules, filter = filter)
+    data_pulled <- srv_init_data("data", data = data)
+    data_validated <- srv_validate_reactive_teal_data(
+      "validate",
+      data = data_pulled,
+      modules = modules,
+      validate_shiny_silent_error = FALSE
+    )
+    data_rv <- reactive({
+      req(inherits(data_validated(), "teal_data"))
+      is_filter_ok <- check_filter_datanames(filter, ls(teal.code::get_env(data_validated())))
+      if (!isTRUE(is_filter_ok)) {
+        showNotification(
+          "Some filters were not applied because of incompatibility with data. Contact app developer.",
+          type = "warning",
+          duration = 10
+        )
+        warning(is_filter_ok)
+      }
+      .add_signature_to_data(data_validated())
+    })
+
+    data_load_status <- reactive({
+      if (inherits(data_pulled(), "teal_data")) {
+        "ok"
+      } else if (inherits(data, "teal_data_module")) {
+        "teal_data_module failed"
+      } else {
+        "external failed"
+      }
+    })
+
     datasets_rv <- if (!isTRUE(attr(filter, "module_specific"))) {
       eventReactive(data_rv(), {
-        if (!inherits(data_rv(), "teal_data")) {
-          stop("data_rv must be teal_data object.")
-        }
+        req(inherits(data_rv(), "teal_data"))
         logger::log_debug("srv_teal@1 initializing FilteredData")
         teal_data_to_filtered_data(data_rv())
       })
     }
 
+    if (inherits(data, "teal_data_module")) {
+      setBookmarkExclude(c("teal_modules-active_tab"))
+      shiny::insertTab(
+        inputId = "teal_modules-active_tab",
+        position = "before",
+        select = TRUE,
+        tabPanel(
+          title = icon("fas fa-database"),
+          value = "teal_data_module",
+          tags$div(
+            ui_init_data(session$ns("data")),
+            ui_validate_reactive_teal_data(session$ns("validate"))
+          )
+        )
+      )
+
+      if (attr(data, "once")) {
+        observeEvent(data_rv(), once = TRUE, {
+          logger::log_debug("srv_teal@2 removing data tab.")
+          # when once = TRUE we pull data once and then remove data tab
+          removeUI(selector = sprintf("#%s a[data-value='teal_data_module']", session$ns("teal_modules-wrapper")))
+          updateTabsetPanel(inputId = "teal_modules-active_tab", selected = names(modules$children)[1])
```

Comment:
I like the change of inserting the `teal_data_module` tab, we can use the same logic to remove it. This will be replaced by `nav_insert()` and `nav_remove()` when we move to `bslib`.
Also, the `updateTabsetPanel` was used because we needed to trigger the selection binding for shiny, since we removed using `removeUI` before, it was not happening.
```suggestion
          shiny::removeTab(
            "teal_modules-active_tab",
            target = "teal_data_module"
          )
```